dft/src/radix_2_bowers.rs

```diff
@@ -110,25 +113,25 @@ fn bowers_g_layer<F: Field>(
     log_half_block_size: usize,
     twiddles: &[F],
 ) {
-    let h = mat.height();
-    let log_block_size = log_half_block_size + 1;
     let half_block_size = 1 << log_half_block_size;
-    let num_blocks = h >> log_block_size;
-
-    // Unroll first iteration with a twiddle factor of 1.
-    for hi in 0..half_block_size {
-        let lo = hi + half_block_size;
-        twiddle_free_butterfly(mat, hi, lo);
-    }
-
-    for (block, &twiddle) in (1..num_blocks).zip(&twiddles[1..]) {
-        let block_start = block << log_block_size;
-        for i in 0..half_block_size {
-            let hi = block_start + i;
-            let lo = hi + half_block_size;
-            dif_butterfly(mat, hi, lo, twiddle);
-        }
-    }
+    let width = mat.width();
+
+    mat.row_chunks_exact_mut(2 * half_block_size)
+        .enumerate()
+        .for_each(|(block, chunks)| {
+            let (hi_chunks, lo_chunks) = chunks.split_at_mut(half_block_size * width);
+            let twiddle = twiddles[block];
+            hi_chunks
+                .par_chunks_exact_mut(width)
```

Comment:
So looks like this is parallelizing over both outer and inner loops? In the future it could be interesting to experiment with other approaches, like parallelizing over the loop with more iterations (based on the current layer's block size). This is nice and simple though, so if the overhead isn't significant then this seems good.

Reply:
Yup, that's precisely it, we parallelize over both outer and inner loops. I am more than happy to experiment with other approaches, as well.